File: R/tagtools/R/mergefields.R
#' Merge the fields of two structures. If there are duplicate fields, the fields in s1 are taken.
#'
#' @param s1 Arbitrary structurese.g., containing metadata or settings.
#' @param s2 Arbitrary structurese.g., containing metadata or settings.
#' @return s A structure containing all of the fields in s1 and s2
#' @export
#' @example s1 <- data.frame( a = 1, b = c(2,3,4))
#'          s2 <- data.frame( b = 3, c = 'cat')
#'		      s <- mergefields(s1,s2)

mergefields <- function(s1,s2) {
  s <- vector(mode = "numeric", length = 0)
  if(missing(s1) | missing(s2)){
    stop("inputs for both s1 and s2 are required")
  }
  
  if (!is.data.frame(s1) & !is.data.frame(s2)){
    stop('Both inputs must be data frames in mergefields\n') 
  }
  
  s <- s2 
  Z <- names(s1) 
  s <- cbind(s1,s) 
  s <- s[, !duplicated(colnames(s))]
  return(s)
}

File: matlab/processing/merge_fields.m
function    s = merge_fields(s1,s2) 
%
%     s = merge_fields(s1,s2)
%     Merge the fields of two structures. If there are duplicate
%     fields, the fields in s1 are taken.
%
%		Inputs:
%		s1, s2 are arbitrary structures e.g., containing metadata or settings.
%
%		Returns:
%		s is a structure containing all of the fields in s1 and s2.
%
%		Examples:
%		s1 = struct('a',1,'b',[2 3 4])
%		s2 = struct('b',3,'c','cat')
%		s = merge_fields(s1,s2)
%		Returns: s containing s.b=[2,3,4], s.c='cat', s.a=1
%
%     Valid: Matlab, Octave
%     markjohnson@st-andrews.ac.uk
%     Last modified: 5 May 2017

s = [] ;
if nargin<2,
   help merge_fields
   return
end

if ~isstruct(s1) & ~isstruct(s2),
	fprintf('Both inputs must be structures in mergefields\n') ;
	return
end
	
s = s2 ;
Z = fieldnames(s1) ;
for k=1:length(Z),
   s = setfield(s,Z{k},getfield(s1,Z{k})) ;
end
